File: skiff/bridge/lib/pi-rpc.js
// bridge/lib/pi-rpc.js
// Zero-dependency client for pi's RPC protocol (`pi --mode rpc`) plus the
// process pool that keeps one live pi per session.
//
// Protocol facts this module leans on (verified against a real pi, and
// documented in pi's rpc.md):
// - Framing is strict JSONL: LF is the only record delimiter. Node's readline
//   also splits on U+2028/U+2029, which are legal inside JSON strings, so the
//   splitter is hand-rolled.
// - Responses and events share stdout; commands carry an `id` and the
//   matching response echoes it. Responses can arrive OUT OF ORDER (a real pi
//   answered a later get_state before an earlier new_session), so correlation
//   is by id, never by arrival order.
// - pi emits extension_ui_request events before any command (fire-and-forget
//   setWidget/setStatus from the orchestrator extension — the live plan and
//   worker readout — plus dialog methods). Dialog methods
//   (select/confirm/input/editor) block the agent until an
//   extension_ui_response arrives, so the bridge auto-cancels them — there is
//   no human on the other end of this bridge, and a deadlocked prompt would
//   wedge every stream subscriber forever.

import { spawn } from "node:child_process";
import { StringDecoder } from "node:string_decoder";
import path from "node:path";
import { mapMessageEntry } from "./session-store.js";

// The overlay entry id for the in-flight assistant message. It is a fixed
// placeholder because the real entry id is only assigned when pi persists the
// message; the stream registry and the message endpoint resolve it to the
// real entry when the file lands (the registry emits a replace at the
// overlay's index; the poll-free fallback — newest assistant message without
// time.completed — keeps working while it streams).
const PENDING_MESSAGE_ID = "<pending>";

const DIALOG_METHODS = new Set(["select", "confirm", "input", "editor"]);

// Read a stream as JSONL records: split on "\n" only, strip one trailing "\r",
// emit each line as a string. Chunks are decoded with StringDecoder so a
// multi-byte UTF-8 sequence split across chunks cannot corrupt a record.
export function createJsonlReader(stream, onLine, onEnd = () => {}, onError = () => {}) {
  const decoder = new StringDecoder("utf8");
  let buffer = "";
  stream.on("data", (chunk) => {
    buffer += decoder.write(chunk);
    let nl;
    while ((nl = buffer.indexOf("\n")) !== -1) {
      let line = buffer.slice(0, nl);
      buffer = buffer.slice(nl + 1);
      if (line.endsWith("\r")) line = line.slice(0, -1);
      onLine(line);
    }
  });
  stream.on("end", () => {
    buffer += decoder.end();
    if (buffer.length > 0) onLine(buffer.endsWith("\r") ? buffer.slice(0, -1) : buffer);
    onEnd();
  });
  stream.on("error", onError);
  return stream;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class PiError extends Error {}

export class PiProcess {
  // sessionFile is null for the bare process spawned by the create flow
  // before pi reports the new session's file; onExit fires once when the
  // child dies (any reason), letting the pool drop its entry. processEvent
  // fires for every parsed stdout event after this process's own handling
  // (so overlay state is current when observers read it) — the stream
  // registry's hook into the live pi stream.
  constructor({ binary, sessionFile = null, cwd, sessionDir, passSessionDir = false, onExit = () => {}, processEvent = null }) {
    this.binary = binary;
    this.sessionFile = sessionFile;
    this.cwd = cwd;
    this.sessionDir = sessionDir;
    this.passSessionDir = passSessionDir;
    this.onExit = onExit;
    this.processEvent = processEvent;
    this.child = null;
    this.exited = false;
    this.nextCommandId = 0;
    this.pending = new Map(); // command id -> { resolve, reject, timer }
    this.busy = false; // agent_start .. agent_settled
    this.pendingMessage = null; // in-flight assistant message assembly
    this.sessionName = null; // set via set_session_name (create flow and rename)
    this.lastOrchestratorActive = null; // latest orchestrator-mode entry_appended
    // Latest fire-and-forget publications from the orchestrator extension
    // (setWidget/setStatus for its key): the live plan/worker readout. The
    // last publication per process is all a poller needs; null means the
    // extension cleared it (mode off, idle).
    this.lastOrchestratorWidget = null; // string[] | null
    this.lastOrchestratorStatus = null; // string | null
    this.lastActivity = Date.now(); // LRU eviction key
    this.stderrTail = "";
    this._exitReason = null;
  }

  start() {
    if (this.child) return;
    // --session-dir is passed only when the bridge's session dir was
    // explicitly overridden. pi's native default is the same directory the
    // bridge scans by default, but organizes files per-cwd (--<cwd>--/
    // buckets) — the layout the pi CLI itself reads and writes. Forcing
    // --session-dir always would make pi write sessions flat into the dir
    // root, invisible to the CLI's project-scoped listings, and silently
    // split the phone's sessions from the CLI's. The explicit-override case
    // accepts the flat layout deliberately: a custom dir has no native pi
    // layout to preserve, and scan and writes must stay in the same place.
    const args = ["--mode", "rpc"];
    if (this.passSessionDir) args.push("--session-dir", this.sessionDir);
    if (this.sessionFile) args.push("--session", this.sessionFile);
    const child = spawn(this.binary, args, { cwd: this.cwd, stdio: ["pipe", "pipe", "pipe"] });
    this.child = child;
    this.exited = false;

    createJsonlReader(child.stdout, (line) => this._onLine(line));
    // stderr is kept as a bounded tail purely for diagnostics on pi failures.
    // Prompt text and the password never appear here (prompts travel stdin).
    const decoder = new StringDecoder("utf8");
    child.stderr.on("data", (chunk) => {
      this.stderrTail = (this.stderrTail + decoder.write(chunk)).slice(-4096);
    });
    child.stderr.on("end", () => {
      this.stderrTail += decoder.end();
    });
    // EPIPE on a dying child is expected; the exit event drives teardown.
    child.stdin.on("error", () => {});
    child.on("error", (err) => this._handleExit(`spawn failed: ${err.code ?? err.message}`));
    child.on("exit", (code, signal) => {
      const reason = code === 0 ? "clean exit" : `exit code ${code}${signal ? ` (${signal})` : ""}`;
      this._handleExit(reason);
    });
  }

  _handleExit(reason) {
    if (this.exited) return;
    this.exited = true;
    this._exitReason = reason;
    for (const { reject, timer } of this.pending.values()) {
      clearTimeout(timer);
      reject(new PiError(`pi process ${reason}`));
    }
    this.pending.clear();
    this.pendingMessage = null;
    this.busy = false;
    this.onExit(this);
  }

  kill() {
    if (this.child && !this.exited) this.child.kill("SIGTERM");
  }

  isBusy() {
    return this.busy;
  }

  getExitReason() {
    return this._exitReason;
  }

  getStderrTail() {
    return this.stderrTail;
  }

  // Send one command and resolve with the correlated response. Rejects when
  // the process dies or the response never arrives (the response timeout is a
  // bound, not a promise that the command failed — pi's prompt response can
  // legitimately trail its events).
  sendCommand(type, fields = {}, { timeoutMs = 30000 } = {}) {
    if (this.exited || !this.child) {
      return Promise.reject(new PiError("pi process is not running"));
    }
    const id = String(++this.nextCommandId);
    const command = { type, id, ...fields };
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new PiError(`pi did not respond to ${type} within ${timeoutMs}ms`));
      }, timeoutMs);
      this.pending.set(id, { resolve, reject, timer });
      this.lastActivity = Date.now();
      try {
        this.child.stdin.write(JSON.stringify(command) + "\n");
      } catch (err) {
        clearTimeout(timer);
        this.pending.delete(id);
        reject(new PiError(`failed to write ${type} command: ${err.message}`));
      }
    });
  }

  _write(obj) {
    if (!this.child || this.exited) return;
    try {
      this.child.stdin.write(JSON.stringify(obj) + "\n");
    } catch {
      // the exit event owns teardown
    }
  }

  _onLine(line) {
    let event;
    try {
      event = JSON.parse(line);
    } catch {
      return; // an unparseable line from pi is not worth surfacing
    }
    if (event.type === "response") this._onResponse(event);
    else {
      switch (event.type) {
        case "message_start":
          this._onMessageStart(event);
          break;
        case "message_update":
          this._onMessageUpdate(event);
          break;
        case "message_end":
          // The entry lands on disk with message_end (verified against real
          // pi), so the overlay's job is done — the file reader takes over.
          this.pendingMessage = null;
          break;
        case "agent_start":
          this.busy = true;
          this.lastActivity = Date.now();
          break;
        case "agent_end":
          if (!event.willRetry) this.busy = false;
          // An assistant message that never reached message_end (aborted or
          // errored run) will never be persisted; keep showing it would freeze
          // the transcript on content that can never stabilize.
          this.pendingMessage = null;
          break;
        case "agent_settled":
          this.busy = false;
          this.pendingMessage = null;
          break;
        case "entry_appended": {
          // The extension API's appendEntry emits this for every custom entry.
          // The orchestrator mode's shared record is such an entry, and before
          // the first assistant message pi buffers it in memory — the file is
          // not flushed until an assistant message exists (verified live), so
          // the process's last seen value is the only record then.
          // newbornSessionObject serves it for the newborn window.
          const entry = event.entry;
          if (entry?.customType === "orchestrator-mode" && typeof entry.data?.active === "boolean") {
            this.lastOrchestratorActive = entry.data.active;
          }
          break;
        }
        case "extension_ui_request":
          this._onExtensionUiRequest(event);
          break;
        default:
          break; // tool_execution_*, queue_update, compaction_*, turn_*, etc.
      }
    }
    if (this.processEvent) this.processEvent(this, event);
  }

  _onResponse(event) {
    const pending = this.pending.get(event.id);
    if (!pending) return; // a response for an already-timed-out command
    clearTimeout(pending.timer);
    this.pending.delete(event.id);
    pending.resolve(event);
  }

  // Only assistant messages stream into the overlay; user/toolResult messages
  // are persisted directly and the file reader surfaces them.
  _onMessageStart(event) {
    const message = event.message ?? {};
    if (message.role !== "assistant") return;
    this.pendingMessage = {
      blocks: new Map(), // contentIndex -> content block under construction
      model: typeof message.model === "string" ? message.model : null,
    };
  }

  // Assemble the in-flight assistant message from deltas. Content blocks are
  // keyed by contentIndex (never by arrival order — toolcalls and text can
  // interleave); toolcall_end carries the authoritative completed call and
  // replaces whatever the deltas accumulated.
  _onMessageUpdate(event) {
    const assembly = this.pendingMessage;
    if (!assembly) return;
    const delta = event.assistantMessageEvent;
    if (!delta) return;
    const index = delta.contentIndex ?? 0;
    switch (delta.type) {
      case "text_start":
        this._block(assembly, index, { type: "text", text: "" });
        break;
      case "text_delta":
        this._block(assembly, index, { type: "text", text: "" }).text += delta.delta ?? "";
        break;
      case "text_end": {
        const block = this._block(assembly, index, { type: "text", text: "" });
        if (typeof delta.content === "string") block.text = delta.content; // authoritative
        break;
      }
      case "thinking_start":
        this._block(assembly, index, { type: "thinking", thinking: "" });
        break;
      case "thinking_delta":
        // The file format's thinking block field is "thinking", not "text" —
        // the assembly must match so mapMessageEntry renders it as reasoning.
        this._block(assembly, index, { type: "thinking", thinking: "" }).thinking += delta.delta ?? "";
        break;
      case "thinking_end": {
        const block = this._block(assembly, index, { type: "thinking", thinking: "" });
        if (typeof delta.content === "string") block.thinking = delta.content;
        break;
      }
      case "toolcall_start": {
        const seed = delta.toolCall ?? {};
        this._block(assembly, index, { type: "toolCall", ...seed });
        break;
      }
      case "toolcall_delta": {
        const block = this._block(assembly, index, { type: "toolCall", id: "", name: "", arguments: "" });
        if (typeof delta.delta === "string") block.arguments = (block.arguments ?? "") + delta.delta;
        break;
      }
      case "toolcall_end":
        if (delta.toolCall) assembly.blocks.set(index, { ...delta.toolCall });
        break;
      default:
        break;
    }
  }

  _block(assembly, index, template) {
    if (!assembly.blocks.has(index)) assembly.blocks.set(index, template);
    return assembly.blocks.get(index);
  }

  // The orchestrator extension publishes its live state fire-and-forget via
  // setWidget/setStatus (see pi's rpc.md extension UI protocol): the widget
  // is the plan/worker readout, the status its compact summary. The bridge
  // captures the last publication per process so skiff's poller can serve
  // it; every other fire-and-forget method stays a display hint and is
  // ignored, and dialog methods are auto-cancelled below.
  _onExtensionUiRequest(event) {
    if (event.method === "setWidget" && event.widgetKey === "orchestrator") {
      this.lastOrchestratorWidget = Array.isArray(event.widgetLines) ? event.widgetLines : null;
      return;
    }
    if (event.method === "setStatus" && event.statusKey === "orchestrator") {
      this.lastOrchestratorStatus = typeof event.statusText === "string" ? event.statusText : null;
      return;
    }
    if (!DIALOG_METHODS.has(event.method)) return; // fire-and-forget: display hints, ignore
    this._write({ type: "extension_ui_response", id: event.id, cancelled: true });
  }

  // The overlay message: file-parsed messages plus this, assembled from the
  // deltas, with no time.completed so skiff's working indicator fires. Blocks
  // are emitted in contentIndex order.
  getPendingMessage() {
    const assembly = this.pendingMessage;
    if (!assembly) return null;
    const blocks = [...assembly.blocks.entries()]
      .sort((a, b) => a[0] - b[0])
      .map(([, block]) => block);
    return mapMessageEntry(
      {
        type: "message",
        id: PENDING_MESSAGE_ID,
        message: { role: "assistant", content: blocks, model: assembly.model },
        timestamp: Date.now(),
      },
      { streaming: true }
    );
  }
}

export class PiPool {
  // processEvent: called for every parsed process event after the process's
  // own handling (see PiProcess); the stream registry's live-event hook.
  // isPinned: a per-sessionFile predicate consulted by LRU eviction — a
  // session with open stream subscribers must not be evicted out from under
  // its own stream (the registry sets both via createBridge's wiring).
  constructor({ binary, sessionDir, sessionDirExplicit = false, defaultCwd, maxProcesses, processEvent = null, isPinned = () => false }) {
    this.binary = binary;
    this.sessionDir = sessionDir;
    this.sessionDirExplicit = sessionDirExplicit;
    this.defaultCwd = defaultCwd;
    this.maxProcesses = maxProcesses;
    this.processEvent = processEvent;
    this.isPinned = isPinned;
    this.processes = new Map(); // sessionFile -> PiProcess (registered, named sessions)
    this.bare = new Set(); // processes mid-create (no session file yet)
  }

  _makeProcess(sessionFile, cwd) {
    const proc = new PiProcess({
      binary: this.binary,
      sessionFile,
      cwd,
      sessionDir: this.sessionDir,
      passSessionDir: this.sessionDirExplicit,
      processEvent: this.processEvent,
      onExit: (p) => this._onProcessExit(p),
    });
    proc.start();
    return proc;
  }

  // Spawn the bare process for the create flow. It is not registered by file
  // until pi reports the new session's file (register()).
  spawnBare(cwd) {
    const proc = this._makeProcess(null, cwd);
    this.bare.add(proc);
    return proc;
  }

  // Move a create-flow process into the keyed pool once it has a session
  // file, then enforce the LRU cap.
  register(proc, sessionFile) {
    this.bare.delete(proc);
    proc.sessionFile = sessionFile;
    this.processes.set(sessionFile, proc);
    this._evict();
  }

  // Return the live process for a session file, spawning (and registering)
  // one on demand. This is the lazy-spawn path for prompt_async/abort.
  async ensure(sessionFile, cwd) {
    const existing = this.processes.get(sessionFile);
    if (existing) return existing;
    const proc = this._makeProcess(sessionFile, cwd);
    this.processes.set(sessionFile, proc);
    this._evict();
    return proc;
  }

  getByFile(sessionFile) {
    return this.processes.get(sessionFile) ?? null;
  }

  // Resolve by session id (file basename without .jsonl) — needed for the
  // newborn-session window when the file does not exist on disk yet.
  getById(id) {
    for (const proc of this.processes.values()) {
      if (proc.sessionFile && path.basename(proc.sessionFile, ".jsonl") === id) return proc;
    }
    return null;
  }

  all() {
    return [...this.processes.values(), ...this.bare];
  }

  _onProcessExit(proc) {
    this.bare.delete(proc);
    if (proc.sessionFile && this.processes.get(proc.sessionFile) === proc) {
      this.processes.delete(proc.sessionFile);
    }
    // Synthetic event so observers (the stream registry) can react to the
    // process vanishing — an in-flight overlay can no longer resolve through
    // deltas once its process is gone.
    if (this.processEvent) this.processEvent(proc, { type: "process_exit" });
  }

  // LRU eviction: kill the oldest idle process until at or under the cap.
  // Busy processes are never evicted (killing a live run would corrupt the
  // session), and pinned processes (open stream subscribers) are never
  // evicted either (their stream depends on the live event hook), so a busy
  // or watched pool may transiently exceed the cap.
  _evict() {
    if (this.processes.size <= this.maxProcesses) return;
    const idle = [...this.processes.values()]
      .filter((proc) => !proc.isBusy() && !this.isPinned(proc.sessionFile))
      .sort((a, b) => a.lastActivity - b.lastActivity);
    while (this.processes.size > this.maxProcesses && idle.length > 0) {
      const victim = idle.shift();
      this.processes.delete(victim.sessionFile);
      victim.kill();
    }
  }

  shutdown() {
    for (const proc of this.all()) proc.kill();
    this.processes.clear();
    this.bare.clear();
  }
}

File: skiff/src/run/pi_rpc.rs
//! A live `pi --mode rpc` child, and the protocol it speaks.
//!
//! Protocol facts this module depends on, all of them load-bearing:
//!
//! - **Framing is strict JSONL, LF only.** Rust's line reader splits on `\n`
//!   and nothing else, which is what the protocol requires. (The Node bridge
//!   had to hand-roll a splitter because `readline` also breaks on U+2028 and
//!   U+2029 — legal characters *inside* a JSON string, so a reply containing
//!   one would corrupt the frame.)
//! - **Responses and events share stdout**, and responses can arrive **out of
//!   order** — a real pi has answered a later `get_state` before an earlier
//!   `new_session`. Correlation is by `id`, never by arrival order.
//! - **Dialog requests block the agent** until answered. There is no human on
//!   the other end of this socket, so they are auto-cancelled by the caller;
//!   a deadlocked prompt would wedge every viewer of that session forever.

use std::collections::HashMap;
use std::path::PathBuf;
use std::process::Stdio;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::time::Duration;

use anyhow::{Context, Result, bail};
use serde_json::{Value, json};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, ChildStdin, Command};
use tokio::sync::{Mutex, mpsc, oneshot};

/// A command's response may legitimately trail its events, so this is a bound
/// on waiting rather than a claim that the command failed.
pub const COMMAND_TIMEOUT: Duration = Duration::from_secs(30);

/// stderr is kept only for diagnostics on a pi that dies. Prompt text never
/// appears there — prompts travel stdin.
const STDERR_TAIL_LIMIT: usize = 4_096;

pub struct PiConfig {
    pub binary: PathBuf,
    pub session_file: PathBuf,
    pub cwd: PathBuf,
    /// Passed as `--session-dir` **only** when skiff's session directory was
    /// explicitly overridden.
    ///
    /// pi's native default is the same directory skiff scans, but organised
    /// into per-cwd buckets — the layout pi's own CLI reads and writes.
    /// Forcing `--session-dir` unconditionally would make pi write flat into
    /// the directory root, invisible to the CLI's project-scoped listings, and
    /// silently split sessions started here from sessions started at a
    /// terminal.
    pub session_dir_override: Option<PathBuf>,
}

pub struct PiProcess {
    stdin: Mutex<ChildStdin>,
    child: Mutex<Child>,
    pending: Arc<Mutex<HashMap<String, oneshot::Sender<Value>>>>,
    next_id: AtomicU64,
    exited: Arc<AtomicBool>,
    stderr_tail: Arc<Mutex<String>>,
}

impl PiProcess {
    /// Spawn pi and start reading its stdout.
    ///
    /// Every event (everything that is not a command response) is forwarded on
    /// the returned receiver. The channel closes when pi exits, which is how
    /// the caller learns the process is gone.
    pub fn spawn(config: &PiConfig) -> Result<(Arc<Self>, mpsc::UnboundedReceiver<Value>)> {
        let mut command = Command::new(&config.binary);
        command.arg("--mode").arg("rpc");
        if let Some(dir) = &config.session_dir_override {
            command.arg("--session-dir").arg(dir);
        }
        command.arg("--session").arg(&config.session_file);
        command
            .current_dir(&config.cwd)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            // Without this a killed skiffd leaves pi children running, holding
            // their session files open.
            .kill_on_drop(true);

        let mut child = command
            .spawn()
            .with_context(|| format!("spawning {}", config.binary.display()))?;

        let stdin = child.stdin.take().context("pi stdin was not piped")?;
        let stdout = child.stdout.take().context("pi stdout was not piped")?;
        let stderr = child.stderr.take().context("pi stderr was not piped")?;

        let pending: Arc<Mutex<HashMap<String, oneshot::Sender<Value>>>> = Arc::default();
        let exited = Arc::new(AtomicBool::new(false));
        let stderr_tail: Arc<Mutex<String>> = Arc::default();
        let (events_tx, events_rx) = mpsc::unbounded_channel();

        tokio::spawn({
            let pending = pending.clone();
            let exited = exited.clone();
            async move {
                let mut lines = BufReader::new(stdout).lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    // An unparseable line from pi is not worth surfacing; the
                    // next frame is independent.
                    let Ok(value) = serde_json::from_str::<Value>(&line) else { continue };
                    if value.get("type").and_then(Value::as_str) == Some("response") {
                        let id = value.get("id").and_then(Value::as_str).unwrap_or_default();
                        if let Some(waiter) = pending.lock().await.remove(id) {
                            // A dropped receiver means the command timed out;
                            // the late response is simply discarded.
                            let _ = waiter.send(value);
                        }
                        continue;
                    }
                    if events_tx.send(value).is_err() {
                        break; // nobody is listening any more
                    }
                }
                // stdout closed: pi is gone. Fail every waiter rather than
                // leaving callers to time out one by one.
                exited.store(true, Ordering::SeqCst);
                pending.lock().await.clear();
            }
        });

        tokio::spawn({
            let stderr_tail = stderr_tail.clone();
            async move {
                let mut lines = BufReader::new(stderr).lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    let mut tail = stderr_tail.lock().await;
                    tail.push_str(&line);
                    tail.push('\n');
                    if tail.len() > STDERR_TAIL_LIMIT {
                        let cut = tail.len() - STDERR_TAIL_LIMIT;
                        // Keep the tail on a character boundary.
                        let cut = (cut..tail.len())
                            .find(|i| tail.is_char_boundary(*i))
                            .unwrap_or(tail.len());
                        *tail = tail[cut..].to_owned();
                    }
                }
            }
        });

        Ok((
            Arc::new(Self {
                stdin: Mutex::new(stdin),
                child: Mutex::new(child),
                pending,
                next_id: AtomicU64::new(0),
                exited,
                stderr_tail,
            }),
            events_rx,
        ))
    }

    pub fn is_alive(&self) -> bool {
        !self.exited.load(Ordering::SeqCst)
    }

    pub async fn stderr_tail(&self) -> String {
        self.stderr_tail.lock().await.clone()
    }

    /// Send one command and wait for the response with the matching id.
    pub async fn command(&self, kind: &str, fields: Value, timeout: Duration) -> Result<Value> {
        if !self.is_alive() {
            bail!("pi is not running");
        }
        let id = (self.next_id.fetch_add(1, Ordering::SeqCst) + 1).to_string();
        let mut frame = json!({ "type": kind, "id": id });
        if let (Some(frame), Value::Object(fields)) = (frame.as_object_mut(), fields) {
            frame.extend(fields);
        }

        let (tx, rx) = oneshot::channel();
        self.pending.lock().await.insert(id.clone(), tx);
        if let Err(err) = self.write(&frame).await {
            self.pending.lock().await.remove(&id);
            return Err(err);
        }

        match tokio::time::timeout(timeout, rx).await {
            Ok(Ok(response)) => Ok(response),
            // The waiter was dropped: pi exited while the command was in
            // flight.
            Ok(Err(_)) => bail!("pi exited before answering {kind}: {}", self.stderr_tail().await),
            Err(_) => {
                self.pending.lock().await.remove(&id);
                bail!("pi did not answer {kind} within {timeout:?}")
            }
        }
    }

    /// Send a frame with no response — the dialog cancellations.
    pub async fn notify(&self, frame: &Value) -> Result<()> {
        self.write(frame).await
    }

    async fn write(&self, frame: &Value) -> Result<()> {
        let mut line = serde_json::to_vec(frame).context("serialising a pi command")?;
        line.push(b'\n');
        let mut stdin = self.stdin.lock().await;
        stdin.write_all(&line).await.context("writing to pi")?;
        stdin.flush().await.context("flushing to pi")?;
        Ok(())
    }

    /// Ask pi to exit. The child is also killed on drop, so this is for the
    /// deliberate case rather than for cleanup.
    pub async fn kill(&self) {
        let _ = self.child.lock().await.start_kill();
    }
}

/// Whether a `extension_ui_request` blocks the agent until answered.
///
/// The fire-and-forget methods are display hints and are ignored; these four
/// wait for a human, and there is none here.
pub fn is_dialog(method: &str) -> bool {
    matches!(method, "select" | "confirm" | "input" | "editor")
}

/// The frame that declines a dialog without answering it.
pub fn cancel_dialog(id: &Value) -> Value {
    json!({ "type": "extension_ui_response", "id": id, "cancelled": true })
}

/// A path to a fake pi for tests: a script that speaks the protocol.
#[cfg(test)]
pub fn fake_pi(dir: &std::path::Path, script: &str) -> PathBuf {
    use std::os::unix::fs::PermissionsExt;
    let path = dir.join("fake-pi");
    std::fs::write(&path, format!("#!/usr/bin/env python3\n{script}")).unwrap();
    std::fs::set_permissions(&path, std::fs::Permissions::from_mode(0o755)).unwrap();
    path
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Reads commands on stdin and answers per a lookup table, so a test can
    /// assert on correlation rather than on timing.
    const ECHO: &str = r#"
import sys, json
for line in sys.stdin:
    cmd = json.loads(line)
    if cmd["type"] == "slow":
        continue                       # never answers
    print(json.dumps({"type": "response", "id": cmd["id"],
                      "success": True, "echo": cmd["type"]}), flush=True)
"#;

    fn config(binary: PathBuf, dir: &std::path::Path) -> PiConfig {
        PiConfig {
            binary,
            session_file: dir.join("s.jsonl"),
            cwd: dir.to_path_buf(),
            session_dir_override: None,
        }
    }

    #[tokio::test]
    async fn a_command_is_answered_by_its_id() {
        let dir = tempfile::tempdir().unwrap();
        let (pi, _events) = PiProcess::spawn(&config(fake_pi(dir.path(), ECHO), dir.path())).unwrap();
        let response = pi.command("prompt", json!({ "message": "hi" }), COMMAND_TIMEOUT).await.unwrap();
        assert_eq!(response["success"], true);
        assert_eq!(response["echo"], "prompt");
    }

    #[tokio::test]
    async fn responses_are_correlated_by_id_not_by_arrival_order() {
        // A real pi has answered a later command before an earlier one.
        const REORDER: &str = r#"
import sys, json
seen = []
for line in sys.stdin:
    seen.append(json.loads(line))
    if len(seen) == 2:
        for cmd in reversed(seen):
            print(json.dumps({"type": "response", "id": cmd["id"], "echo": cmd["type"]}), flush=True)
"#;
        let dir = tempfile::tempdir().unwrap();
        let (pi, _events) =
            PiProcess::spawn(&config(fake_pi(dir.path(), REORDER), dir.path())).unwrap();
        let first = pi.command("first", json!({}), COMMAND_TIMEOUT);
        let second = pi.command("second", json!({}), COMMAND_TIMEOUT);
        let (first, second) = tokio::join!(first, second);
        assert_eq!(first.unwrap()["echo"], "first", "each caller got its own answer");
        assert_eq!(second.unwrap()["echo"], "second");
    }

    #[tokio::test]
    async fn events_reach_the_listener_and_responses_do_not() {
        const MIXED: &str = r#"
import sys, json
print(json.dumps({"type": "agent_start"}), flush=True)
for line in sys.stdin:
    cmd = json.loads(line)
    print(json.dumps({"type": "message_start"}), flush=True)
    print(json.dumps({"type": "response", "id": cmd["id"], "success": True}), flush=True)
"#;
        let dir = tempfile::tempdir().unwrap();
        let (pi, mut events) =
            PiProcess::spawn(&config(fake_pi(dir.path(), MIXED), dir.path())).unwrap();
        assert_eq!(events.recv().await.unwrap()["type"], "agent_start");
        pi.command("prompt", json!({}), COMMAND_TIMEOUT).await.unwrap();
        assert_eq!(events.recv().await.unwrap()["type"], "message_start");
    }

    #[tokio::test]
    async fn a_command_that_is_never_answered_times_out_without_wedging_the_next() {
        let dir = tempfile::tempdir().unwrap();
        let (pi, _events) = PiProcess::spawn(&config(fake_pi(dir.path(), ECHO), dir.path())).unwrap();
        let err = pi.command("slow", json!({}), Duration::from_millis(150)).await.unwrap_err();
        assert!(err.to_string().contains("did not answer"), "{err}");
        // The timed-out waiter must not block later commands.
        let ok = pi.command("prompt", json!({}), COMMAND_TIMEOUT).await.unwrap();
        assert_eq!(ok["echo"], "prompt");
    }

    #[tokio::test]
    async fn a_dying_pi_fails_its_waiters_rather_than_making_them_wait() {
        const DIE: &str = "import sys\nsys.exit(3)\n";
        let dir = tempfile::tempdir().unwrap();
        let (pi, _events) = PiProcess::spawn(&config(fake_pi(dir.path(), DIE), dir.path())).unwrap();
        // Long timeout: the point is that the failure arrives immediately.
        let err = pi.command("prompt", json!({}), Duration::from_secs(30)).await.unwrap_err();
        assert!(err.to_string().contains("exited"), "{err}");
        assert!(!pi.is_alive());
    }

    #[tokio::test]
    async fn an_unparseable_line_does_not_break_the_frames_around_it() {
        const NOISE: &str = r#"
import sys, json
print("not json at all", flush=True)
print(json.dumps({"type": "agent_start"}), flush=True)
for line in sys.stdin:
    cmd = json.loads(line)
    print(json.dumps({"type": "response", "id": cmd["id"], "ok": 1}), flush=True)
"#;
        let dir = tempfile::tempdir().unwrap();
        let (pi, mut events) =
            PiProcess::spawn(&config(fake_pi(dir.path(), NOISE), dir.path())).unwrap();
        assert_eq!(events.recv().await.unwrap()["type"], "agent_start");
        assert_eq!(pi.command("x", json!({}), COMMAND_TIMEOUT).await.unwrap()["ok"], 1);
    }

    #[tokio::test]
    async fn a_line_separator_inside_a_string_does_not_split_the_frame() {
        // U+2028 is legal inside a JSON string. Node's readline splits on it,
        // which is why the bridge had to hand-roll a splitter; the Rust line
        // reader breaks on \n alone, so this is correct for free.
        const SEP: &str = r#"
import sys, json
print(json.dumps({"type": "agent_start", "note": "a b"}), flush=True)
sys.stdin.readline()
"#;
        let dir = tempfile::tempdir().unwrap();
        let (_pi, mut events) =
            PiProcess::spawn(&config(fake_pi(dir.path(), SEP), dir.path())).unwrap();
        let event = events.recv().await.unwrap();
        assert_eq!(event["note"], "a\u{2028}b", "the frame survived intact");
    }

    #[test]
    fn dialog_methods_are_the_ones_that_block_the_agent() {
        for method in ["select", "confirm", "input", "editor"] {
            assert!(is_dialog(method));
        }
        for method in ["setWidget", "setStatus", "notify"] {
            assert!(!is_dialog(method), "{method} is fire-and-forget");
        }
    }
}
